### DISCUSSION TITLE
Can the 256KB file size limit be increased to 1MB?

### DISCUSSION THREAD

**Discussion by thosakwe** [Shopify/function-examples#425] [Feedback: General Improvements]
Hey Shopify Team,

Shopify Functions is fantastic, but, I've hit a roadblock that I believe is affecting not just me but several other developers out there.

The 256KB file size limit for Shopify Functions is a bit tight, especially when working with Rust.

Public apps often have to cover a wide variety of use cases and configurations. This presents a challenge, because if you have many configuration options, the `serde` crate will generate a TON of monomorphized code, which bloats the executable size.

For example, [see this previous discussion about the 11M instruction count limit](https://github.com/Shopify/function-examples/discussions/329). My TypeScript Function keeps hitting the instruction count limit, but the file size limit is preventing me from releasing a more efficient Rust version that executes fewer instructions.

My suggestion: how about boosting the file size limit to `1 MB`?

This should give developers more breathing room to create robust Shopify Functions without constantly worrying about fitting into a tight space.

I understand the need for limitations, but the current constraints seem to hinder the potential of what Shopify Functions can achieve.

I'm curious to know if others in the community share similar experiences or if there are alternative suggestions.

Looking forward to your thoughts and hopefully an expanded file size limit soon. 🙌

Best,
Tobe

**Comment by davejcameron** [comment]
We have done some significant work on improving the Rust binary size, especially with shopify_functions 0.8.0. Let me know if you are still encountering issues with the file size limit in Rust.

**Reply by nellodb** [reply]
Hi, I'm no Rust developer but as soon as I tried the Typescript road I encountered the same issues thosakwe ran into.

Our use-case envisages JWT verifications and while I'd love to get access to Shopify Network through functions, Enterprise is nowhere near our budget. I went through the tips in the https://shopify.dev/docs/apps/build/functions/programming-languages/rust-for-functions#binary-size-tips but hit the aforementioned limit pretty much immediately after installing `jwt-simple = "0.12.10"` and using the barebone necessities with two attributes and a verification attempt.

I'm certain a a 1 MB boost or even 512k would go a long way.

Thanks

**Reply by Sam115200** [reply]
> We have done some significant work on improving the Rust binary size, especially with shopify_functions 0.8.0. Let me know if you are still encountering issues with the file size limit in Rust.

I can verify that, just by updating to shopify_functions 0.8.0, I've seen, on average, around 20KB+ binary reduction (using rust). Thank you to all the developers who made this possible, this gives us some room to breathe.

**Reply by thosakwe** [reply]
I am no longer experiencing these issues. The main contributor to binary size was the `serde` Rust library. I totally understand why `serde` produces so much code, but it's not suitable for serializing/deserializing complex structures in a WebAssembly context, where you need the smallest possible binaries.

My solution was to create a custom script that generates deserialization code. My app still uses `serde_json` for the actual parsing of JSON to an abstract syntax tree, but the deserialization is a ~3K line, auto-generated file that uses minimal monomorphization and produces a far smaller file. With serde, the binary was over 400KB.

**Comment by igorwulff** [comment]
I've been working with Tinygo and keeping things under 256kb is a real challenge. It would help a lot if this gets increased to 1024kb or even 512kb. I of course don't fully understand the architecture behind Shopify Functions internally, and the reasons for having these low limits, but I assume increasing this file size won't drastically decrease performance. Am I right? If that is the case I do advocate for a more developer friendly configuration, with usage also embracing (incl. documentation+examples) languages like TinyGo and Zig with more then just a small reference in the WASM part of the docs.

The problem with this low limit and I see it also with comments above here, is that utilizing certain dependencies for serialization or other things, quickly add up a lot. For json encoding and decoding I now utilized an rudimentary custom method for dealing with this. But even utilizing the "fmt" package does add up.

For tinygo my recommended built arguments at this time are:
```tinygo build -o cart-total-check.wasm -target=wasi -panic trap -no-debug -scheduler=none -opt=z main.go```

Which does reduce my simple function to 135kb and includes imports of "fmt" and "os", but I had to avoid using packages that are really helpful and development friendly, had to dive into tinygo build optimizations (disable scheduler, no-debug, ...), and I need to be cautious about adding any code and dependencies to it, since it might go beyond the limit very quickly.

Next to this; I'm very surprised with the way Shopify is embracing Rust, while the natural users of Shopify are still often clients and developers not familiar with any complex compiled languages. I would advocate and advice having more documentation and support for alternatives like tinygo and zig, also because Go especially is a much easier language to utilize then Rust for most.
